Tools/Scripts/libraries/reporelaypy/reporelaypy/checkoutroute.py

```diff
@@ -0,0 +1,189 @@
+# Copyright (C) 2021 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import re
+
+from flask import abort, current_app, json as fjson, redirect, Flask, Response
+from reporelaypy import Database
+from webkitflaskpy import AuthedBlueprint
+from webkitscmpy import Commit, remote
+
+
+class Redirector(object):
+    class Encoder(json.JSONEncoder):
+        def default(self, obj):
+            if isinstance(obj, dict):
+                return {key: self.default(value) for key, value in obj.items()}
+            if isinstance(obj, list):
+                return [self.default(value) for value in obj]
+            if not isinstance(obj, Redirector):
+                return super(Checkout.Encoder, self).default(obj)
+
+            return dict(
+                name=obj.name,
+                url=obj.url,
+            )
+
+    @classmethod
+    def bitbucket_generator(cls, base):
+        def redirector(commit):
+            if commit and commit.hash:
+                return redirect('{}/commits/{}'.format(base, commit.hash))
+            return redirect('{}/commits'.format(base))
+
+        return redirector
+
+    @classmethod
+    def trac_generator(cls, base):
+        def redirector(commit):
+            if commit and commit.revision:
+                return redirect('{}/changeset/{}/webkit'.format(base, commit.revision))
+            return redirect(base)
+
+        return redirector
+
+    @classmethod
+    def github_generator(cls, base):
```

Comment:
This seems identical to `bitbucket_generator`, do you think we should unify them?

Reply:
They're _almost_ identical:
<pre>
@classmethod
    def bitbucket_generator(cls, base):
        def redirector(commit):
            if commit and commit.hash:
                return redirect('{}/commits/{}'.format(base, commit.hash))
            return redirect('{}/commits'.format(base))

        return redirector
</pre>
<pre>
@classmethod
    def github_generator(cls, base):
        def redirector(commit):
            if commit and commit.hash:
                return redirect('{}/commit/{}'.format(base, commit.hash))
            return redirect('{}/commits'.format(base))

        return redirector
</pre>
BitBucket uses `commits/<hash>` and GitHub uses `commit/<hash>`

Reply:
I see.